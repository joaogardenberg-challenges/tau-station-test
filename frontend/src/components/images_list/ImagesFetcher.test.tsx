import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import range from 'lodash/range'
import ThemeProvider from 'providers/Theme'

const useDispatch = jest.fn()
const useSelector = jest.fn()
const useOnScreen = jest.fn()
const fetchImages = jest.fn()

jest.doMock('react-redux', () => ({ useDispatch, useSelector }))
jest.doMock('hooks/useOnScreen', () => useOnScreen)
jest.doMock('actions', () => ({ fetchImages }))

const ImagesFetcher = require('./ImagesFetcher').default

const renderWithTheme = (ui: any, options: any = {}) =>
  render(ui, { wrapper: ThemeProvider, ...options })

describe('Image Filler', () => {
  const lastId = 21

  const props = {
    perRow: 5,
    setMissingEnd: jest.fn()
  }

  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn())
    useSelector.mockReturnValue(lastId)
  })

  it('renders successfully when visible', () => {
    useOnScreen.mockReturnValue(true)
    const { container } = renderWithTheme(<ImagesFetcher {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('renders successfully when not visible', () => {
    useOnScreen.mockReturnValue(false)
    const { container } = renderWithTheme(<ImagesFetcher {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('fetches the images when visible', () => {
    useOnScreen.mockReturnValue(true)
    renderWithTheme(<ImagesFetcher {...props} />)

    expect(fetchImages).toHaveBeenCalledTimes(1)
    expect(fetchImages.mock.calls[0][0]).toEqual({
      fromId: lastId - 1,
      limit: props.perRow * 10
    })
  })

  it('sets missing end correctly', () => {
    const missingEnd = 3
    useOnScreen.mockReturnValue(true)
    fetchImages.mockImplementation((_, onSuccess) =>
      onSuccess(range(props.perRow * 10 - missingEnd))
    )

    renderWithTheme(<ImagesFetcher {...props} />)

    expect(props.setMissingEnd).toHaveBeenCalledTimes(1)
    expect(props.setMissingEnd).toHaveBeenCalledWith(missingEnd)
  })

  it('only fetches the images when turning visible', () => {
    useOnScreen.mockReturnValue(false)
    const { rerender } = renderWithTheme(<ImagesFetcher {...props} />)
    expect(fetchImages).not.toHaveBeenCalled()

    useOnScreen.mockReturnValue(true)
    rerender(<ImagesFetcher />)
    expect(fetchImages).toHaveBeenCalledTimes(1)
  })

  it('fetches the images on button click', () => {
    useOnScreen.mockReturnValue(false)
    renderWithTheme(<ImagesFetcher {...props} />)

    const button = screen.getByRole('button', { name: 'Load more' })
    userEvent.click(button)

    expect(fetchImages).toHaveBeenCalledTimes(1)
  })
})
