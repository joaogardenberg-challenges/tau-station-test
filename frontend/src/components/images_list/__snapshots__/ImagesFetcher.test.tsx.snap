// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Filler renders successfully when not visible 1`] = `
<div>
  <div
    class="sc-bdfBQB iQTOWu"
  >
    <button
      class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-1hf73u2-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Load more
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`Image Filler renders successfully when visible 1`] = `
<div>
  <div
    class="sc-bdfBQB iQTOWu"
  >
    <span
      class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-lepzwj-MuiCircularProgress-root"
      role="progressbar"
      style="width: 45px; height: 45px;"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="19.5"
          stroke-width="5"
        />
      </svg>
    </span>
  </div>
</div>
`;
